### DISCUSSION TITLE
Making multiple VG Plots from large parquet file

### DISCUSSION THREAD

**Discussion by bbollen23** [uwdata/mosaic#557] [Q&A]
Hi there,
So far, Mosaic has been great for us. But I'm currently having trouble getting the correct paradigm down for how to interact with Mosaic in the "correct" way.

The situation is that our application loads several large files when the application opens. Our idea was that if we can use "vg.loadParquet" to initialize the table, we can generate several different plots from this large table.

But I'm a bit confused on how to go about this. I've had some success with creating temp tables using coordinator.exec() and just filtering what I want out of the large table with a raw sql query, but this seems to go against what Mosaic is trying to do. My other guess was that each chart is essentially the same chart except with different data automatically filtered out.

So I guess all I'm asking is 
1) given a table that is already generated in DuckDB, how can I generate several charts from this one dataset with varying "where" clauses
2) Is initializing a table before chart creation a smart way to go about generating charts or is this an anti-pattern in Mosaic's case?

Any help would be greatly appreciated. This may be a very simple question, but I've scoured the documentation and can't find any that seems to match what I'm trying to do. Thanks!

**Comment by domoritz** [comment]
I'm not sure I totally follow. Mosaic works best when you operate on a single dataset so that cross-filtering etc works across the various clients. Can you give a bit more detail on the use case? Do you want to have a a single source dataset (combining multiple parquet files) and then show different subsets on different charts?

**Reply by bbollen23** [reply]
Thank you for the quick response! Of course. Here's the basic idea. We load one file into duckDb.

```js
 await vg.coordinator().exec([vg.loadParquet(tableName, url)]);
 ```

Now, like you said, we basically want to show subsets of this data in various charts. For right now, I'm considering static visualizations. Each of the contents of each of these charts can be described as some SQL query. For example,

Chart1: "SELECT * FROM 'tableName' WHERE location = 1" 
Chart 2: "SELECT col1, avg(col2) FROM 'tableName' GROUP BY col1 "
Chart 3: "SELECT col1 FROM 'tableName' WHERE col4 = 5 AND col2 = 3"

So each of these would return some dataset that I want to plot in various ways. I just can't seem to understand how I can define a VG plot that pulls from the filtered data (i.e. is operating on the results of these queries). Every example that I've seen that has a filtering on a large table like this usually does the filtering during loading of the parquet file like such:

```js
await vg.coordinator().exec([
  vg.loadParquet("athletesBatched", "data/athletes.parquet", {
  select: ["*", "10 * CEIL(ROW_NUMBER() OVER (PARTITION BY sport) / 10) AS batch"],
  where: "height IS NOT NULL"
})
]);
```

This seemed odd for our use case since when I'm creating all these charts, my parquet file has already been initialized in DuckDB.

**Reply by domoritz** [reply]
Interesting use case and should be supported via a custom [`query`](https://idl.uw.edu/mosaic/core/#clients) implementation that combines your custom query with the query for the mark. I think we don't have a way right now for a vgplot mark to specify an additional query but maybe @jheer can think of another way.

**Reply by bbollen23** [reply]
Well we didn't necessarily want to change the marks, we just wanted to render a lot of different charts from the same dataset. But we found out a way to get it at least working, although it did still seem odd.

For context, we are using Vue. In this use case, we are rendering a grid of charts. The axes of the grid itself are two parameters. We want to display a chart in the coordinate (X,Y) where "col_1 = X" and "col_2 = Y". But wanted no other interaction with the charts themselves. Then, the user can change what columns map to the X axis and Y axis (outside of the chart, in a wrapper component that is controlled separately), so the predicates end up being extremely variable. 
 
 Here, we basically made a "tagSelection" variable which mimics a user selection. The predicate here would then be based on the input to the component. Below is a single chart that would be rendered. If there's a better way to go about this, I'd love to know.

```js
        //Create a vg selection
        const tagSelection = computed(() => vg.Selection.single());
        // Set a unique source so we do not chain filters
        const source = 'test_source';
        // Create clause with filter predicate
        const clause = { source, predicate: "col1 = X AND col2 = Y" };
        // Update selection
        tagSelection.value.update(clause);


        // Creates chart, filtered by the selection that uses the query.
        const chart = vg.plot(
            // Filter based on selection
            vg.lineY(
                vg.from(`my_big_table`,{
                    filterBy:tagSelection.value
                }),
                {
                    x: xAxisName,
                    y: yAxisName,
                    stroke: lineColor,
                    strokeWidth: strokeWidth,
                }
            ),
            // General settings.
            vg.width(width),
            vg.height(height),

            // Gets rid of axes and margins!
            vg.axis(false),
            vg.margin(0)
        );
```

My assumption is that since this source has been initialized, we can still filter on the individual charts via a human selection, correct? So if a user wants to brush over one of these charts, if all the sources are unique, they should be able to filter additional data from that individual chart?

We plan on using this same table for other aggregate charts as well, so its better for us to have one source of truth and a common way to generate the aggregations. Thank you for your help!

**Reply by domoritz** [reply]
I'm not totally following but my suggestions was to implement a custom client and generate the appropriate query in the `query` method. I'm not sure how well we support the approach you have here of wrapping vgplot.

**Reply by Dev-Lan** [reply]
Just to chime in :).

This is an image of the visualization we are trying to create. From this paper: https://vdl.sci.utah.edu/publications/2021_vis_loon/
But this style vis is also in a few other papers (MulteeSum: A Tool for Comparative Spatial and Temporal Gene
Expression Data, Pathline: A Tool For Comparative Functional Genomics).

As brian already mentioned, all of this data comes from one large table, and each line is aggregating a large subset of the data. The subsets do not overlap between the charts.

![image](https://github.com/user-attachments/assets/39a2c444-471b-418c-a372-6688a60bc878)

**Reply by jheer** [reply]
Hi all, I would want to implement this by loading the main table once, and creating a grid of separate vgplot `plot` instances. Sounds like this is what you've initially tried to do.

As I understand it, the "tricky" bit is applying different fixed filtering criteria to these different plots. Right now all filtering criteria pushed to the database are passed in via the `filterBy` selection. We could add an additional data property for marks (maybe named `where`?) which includes a "hardwired" SQL filtering predicate that is always applied to that plot.

Do you think this would address your use case? It's certainly something I've seen requested in the past.

**Reply by domoritz** [reply]
Oh, so it's a facet/small multiple charts. We have that in https://idl.uw.edu/mosaic/examples/walmart-openings.html so you can look at how we do that there. Do you want to make this whole facet one mosaic client or each small chart a client?

**Reply by bbollen23** [reply]
> Hi all, I would want to implement this by loading the main table once, and creating a grid of separate vgplot `plot` instances. Sounds like this is what you've initially tried to do.
> 
> As I understand it, the "tricky" bit is applying different fixed filtering criteria to these different plots. Right now all filtering criteria pushed to the database are passed in via the `filterBy` selection. We could add an additional data property for marks (maybe named `where`?) which includes a "hardwired" SQL filtering predicate that is always applied to that plot.
> 
> Do you think this would address your use case? It's certainly something I've seen requested in the past.

Yes! Thats what I was looking for. A lot of our charts do require user filtering, but they essentially "begin" as an already filtered dataset. So when working with Vue and React, its great to be able to pass these in as parameters to our component which can be passed to the plot for filtering. 

Just like you said, our solution was still to generate separate VG plots (and we made our grid as just a Vue component which loops to create all the charts). Our way of doing that "fixed filtering" was to mimic an initial selection on the data.

**Reply by bbollen23** [reply]
> Oh, so it's a facet/small multiple charts. We have that in https://idl.uw.edu/mosaic/examples/walmart-openings.html so you can look at how we do that there. Do you want to make this whole facet one mosaic client or each small chart a client?

To be honest, I'm not really clear on what the term "client" really means, but I can explain our desired workflow.

User selects an "experiment". For one experiment, one file is loaded using "vg.loadParquet()". This file contains all the information we need for the experiment.

For a single experiment, there are several different views, each rendering one or more individual plots (each plot would be one call of "vg.plot()"). Our eventual goal is to be able to link all these plots together. So, we may have a bar chart which shows the distribution of the masses of cells. When a user brushes over the bar chart, several other plots will be updated to show the filtering that the user applied.

In the view that @Dev-Lan showed, our goal is that a user can click on a single plot container (so not actually brushing on the vg.plot itself -- they are generally static) and by doing so, it would update a central store which says "Since the user clicked on this plot, we need to update all the filter by statements for all other Vg.plots". 

My guess is that based on the documentation and based on the example we gave above, this is possible as long as there is a central truth of what the "filterBy" statements are applying to -- the "source". For example, if all vg.plots have the following filter by statement:

```js
let filterBy = {
   source: "central_truth_source",
   predicate: "my_predicate"
}
```

Then each of the plots, regardless of where they are should be filtered correctly. Then, we can apply additional filtering if necessary to the individual plots (using something like vg.Selection.intersect() ).

Does that make sense?

**Reply by domoritz** [reply]
> To be honest, I'm not really clear on what the term "client" really means

A client in Mosaic is this: https://idl.uw.edu/mosaic/core/#clients
